integration/container/pause_test.go

```diff
@@ -62,7 +63,7 @@ func TestPauseFailsOnWindowsServerContainers(t *testing.T) {
 	poll.WaitOn(t, container.IsInState(ctx, client, cID, "running"), poll.WithDelay(100*time.Millisecond))
 
 	err := client.ContainerPause(ctx, cID)
-	assert.Check(t, is.ErrorContains(err, "cannot pause Windows Server Containers"))
+	assert.Check(t, is.ErrorContains(err, containerderrdefs.ErrNotImplemented.Error()))
```

Comment:
I think `errors.Is(err, containerderrdefs.ErrNotImplemented)` would be better than `is.ErrorContains(err, containerderrdefs.ErrNotImplemented.Error())`, to test for the (possibly wrapped) error type rather than a stringifictation match.

Reply:
Afaiu `errors.Is` does not work here because actual error is: `Error response from daemon: Cannot pause container <container id>: not implemented`. That why original logic have been using "contains" instead of is".

Look https://github.com/moby/moby/pull/41479#issuecomment-787670106

Reply:
`errors.Is` tests the _structure_ of the error, not the stringification. So unless the error has lost its type somewhere between where it was raised and here (which won't be the case for "local" containerd, but perhaps something in the "remote" containerd variant has stripped this before we get it?) then it can be examined with `errors.Is` to see if it is, _or wraps_ a particular error. See [Working with Errors in Go 1.13](https://blog.golang.org/go1.13-errors#TOC_3.2.) for details. It's this behaviour that `containerderrdefs.IsNotImplemented` wraps, if you want to be isolated from the error-handling machinery and just use `containerderrdefs` APIs.

The [original introduction of `containerderrdefs` package](https://github.com/containerd/containerd/commit/a4fadc596b219ad75837d97a1be3bc40e06d97ce) suggests that the type is _intended_ to be preserved across gRPC boundaries, so if `errors.Is` does not work here, I suspect that might be a bug in containerd.

Perhaps you're confusing `errors.Is` with `is.Error`. The latter tests for stringification-equality, and, like `is.ErrorContains` is susceptible to random string changes as we saw here. I suspect the reason for this failure is that containerd or hcsshim changed their error string in this case, while this project wasn't running containerd tests on CI.

Reply:
Looks that either I understand your proposal incorrectly or this logic just does not work:
```
[2021-03-25T18:01:04.722Z] === FAIL: github.com/docker/docker/integration/container TestPauseFailsOnWindowsServerContainers (1.82s)
[2021-03-25T18:01:04.722Z]     pause_test.go:67: assertion failed: expression is false: errors.Is(err, containerderrdefs.ErrNotImplemented)
```

Reply:
Hmm, that's disappointing. My guess is that it's actually Docker turning the `containerderrdefs.ErrNotImplemented` into some other error as it passes through API layers.

Ah! I forgot that we're going through the Docker REST API here, so we _should_ be testing [`client.IsErrNotImplemented(err)`](https://github.com/moby/moby/blob/797b974cb90e32c7efe9e324d7459122c3f6b7b0/client/errors.go#L121-L129), not `errors.Is(err, containerderrdefs.ErrNotImplemented)`, because the various error-code sources on the server side are folded down to the Docker client types [when they come out of the REST call](https://github.com/moby/moby/blob/797b974cb90e32c7efe9e324d7459122c3f6b7b0/client/request.go#L118).

Reply:
Ok. Switched back to that one.